### DISCUSSION TITLE
Point cloud classification/regression with 400.0-400K points per input

### DISCUSSION THREAD

**Discussion by davodogster** [pyg-team/pytorch_geometric#4834] [Q&A]
Hi I successfully wrote the PyG dataloader to iteratively read in points clouds ranging from 400 points to 400,000 points (but most have >100K points!). I have X,Y,Z coords + ReturnIntensity. number of points clouds = 1000-10000. I have multiple parameters (multi-task regression) to estimate per point cloud. I will start with just single paramater regression. Which architecture/network would you recommend??? 
Any advice is much appreciated. And something that can learn both global and local features. The network needs to be quite computationally efficient and able to train on my 3090 GPU :)

I will probably just use the KNNGraph function to convert the PCDs into graphs (on the fly?).

Thanks everyone!! Sam

**Comment by rusty1s** [comment]
I would recommend to start small and then improve gradually. Importantly, I think you should pre-process your dataset to have a uniform size, e.g., to slice big point clouds into smaller chunks and save them as individual examples. You can then start by applying `PointConv` operators on a pre-defined `kNN`-graph, and apply some global pooling at the end. One can then extend the network to incorporate features such as hierarchical pooling, etc (similar to our `examples/pointnet2_*.py` examples).

**Reply by davodogster** [reply]
Hi @rusty1s thanks for your advice! Okay I will consider that. Each point cloud represents a single object though, so it would be breaking up objects into chunks?
I should also note that the bigger point clouds don't actually occupy much more space in X,Y,Z, they are just much more dense/detailed objects.

By pre-defined kNN-graph do you mean save PCDs in memory as graphs, instead of converting them to graphs on the fly for every batch?

**Reply by rusty1s** [reply]
As an alternative, you can just sample points to make them less dense.

> By pre-defined kNN-graph do you mean save PCDs in memory as graphs, instead of converting them to graphs on the fly for every batch?

Yes, saving the graph in a pre-processing step will save you some runtime :)

**Reply by davodogster** [reply]
Ok but I'll have to chose a good value for K.. do you have any recommendation on doing that?

**Reply by rusty1s** [reply]
This is a hyperparameter and needs to be verified experimental. I think `k=10` or `k=20` is a good starting point.

**Reply by davodogster** [reply]
Hi @rusty1s Do you think "Scaling Up GNNs via Remote Backends" (looks like a cool new feature!) https://github.com/pyg-team/pytorch_geometric/discussions/4834 will be useful so I can train on Graphs with 500K-2M edges? My Graphs (from tree point clouds) have varying number of points so each batch will have different n_edges.

**Reply by rusty1s** [reply]
Yes, but 2M edges doesn't actually sound that big as long as you use `NeighborLoader`.

**Reply by davodogster** [reply]
@rusty1s Ok Thanks, Im doing Graph (multitask) regression. So do I just replace Dataloader with NeighborLoader?
![image](https://user-images.githubusercontent.com/46079516/212792334-3e6b1c71-7273-4660-a2ba-349a7b2b6854.png)

And Im using T.KNNGraph to compute the edges and store them in memory
![image](https://user-images.githubusercontent.com/46079516/212792480-54ce0fdc-4ebc-4862-a3f5-1724c12b6bf1.png)

**Reply by rusty1s** [reply]
If you want to predict graph-level properties, then you should keep the `DataLoader`, but you may need to set a low batch size for 60k nodes, e.g., `batch_size=1` or `batch_size=2`.

**Reply by davodogster** [reply]
@rusty1s Thanks, and is this the correct way to do multitask learning - 3x regression tasks:

![image](https://user-images.githubusercontent.com/46079516/216458795-12a04b5b-b9de-4584-a9c2-05e589a0c366.png)

![image](https://user-images.githubusercontent.com/46079516/216458913-54488b52-7c03-4cc3-ba30-11f0daae5ed8.png)

Regards, Sam

**Reply by rusty1s** [reply]
Yes, this looks correct. You can also consider computing the output at once, e.g., via `self.regressor = Linear(FILTERS, 3)`.

**Reply by davodogster** [reply]
@rusty1s Thanks for checking it! 

If I use self.regressor = Linear(FILTERS, 3) will it share the same regressor for all tasks instead of a unique regressor for each? Would there be an accuracy difference or it has the same functionality as using 3x unique Linear layers?

**Reply by rusty1s** [reply]
It has the same functionality, it is just more efficient.

**Reply by davodogster** [reply]
@rusty1s Thanks! Do you know how I can add more features to train the model? Currently I'm just using pos (X,Y,Z) and edge_index as features but I want to add 1-2 more features per position (coordinate).

![image](https://user-images.githubusercontent.com/46079516/222037178-b07fba33-3618-47db-a494-992316581850.png)

**Reply by rusty1s** [reply]
That should be doable by just increasing `in_channels` to your desired number of input features.

**Reply by davodogster** [reply]
Thanks for the advice! Here is my data with the new feature "I". Edges are computed as a pre-transform with KnnGraph.

![image](https://user-images.githubusercontent.com/46079516/222307500-3e8f50ff-6f72-4097-a58e-bfdba34b6867.png)

`        logits = model(data.I.to(device), data.pos.to(device), data.edge_index.to(device), data.batch.to(device)) `

I got it working by changing 3 input features (X,Y,Z) to 1 input feature (I)
`self.conv1 = PointNetLayer(1, FILTERS)`

**Reply by davodogster** [reply]
@rusty1s Do you think it will it improve performance if I use 4 features = X,Y,Z,I, pos= X,Y,Z vs 1x features = I, pos = X,Y,Z ?

**Reply by rusty1s** [reply]
This depends, best to try out. Moving the `pos` to the input features means that not only the relations between points, but also their absolute values will be used within the model. That means that your model is no longer translation-invariant.

**Reply by davodogster** [reply]
Thanks @rusty1s I'll try that out!
 
I now have three parameters to estimate in multitask learning. One of them has 4 states but they are continuous values 3,4,5,6. The other params range from 0.2,1.0 (10 states) and 30,90. Do you know how I can adjust or weight the loss in multitask learning to account for these different scales and parameter ranges?

Here is my current code for the losses. It assumes equal weighting for all params.

![image](https://user-images.githubusercontent.com/46079516/223896209-1375d2a3-e276-498b-adba-4a492231395f.png)

**Reply by davodogster** [reply]
> It has the same functionality, it is just more efficient.

Thanks, I have made the change and hopefully I am doing it correctly:
I also changed:
![image](https://user-images.githubusercontent.com/46079516/223910908-6c75e1ca-ba4d-44d2-9898-8f23ad1f7930.png)

to:
![image](https://user-images.githubusercontent.com/46079516/223910721-f32a62b2-2813-49c6-8f3a-ff3739395217.png)
![image](https://user-images.githubusercontent.com/46079516/223910765-cfb77fc8-5b3b-4b0f-81ee-bb664e505b07.png)

**Reply by rusty1s** [reply]
I think you would want to normalize the labels individually, e.g., via standard normalization. In this case, you do not need to perform any weighting of the loss.

**Comment by davodogster** [comment]
Thanks, so if I perform min-max scale all labels (3) to 0-100 will that suffice? MinMaxScaler

**Reply by rusty1s** [reply]
Yes, that works as well.

**Reply by davodogster** [reply]
@rusty1s Is it possible in PyG to perform GNN Explainability for my MLP Pointnet GNN for graph regression? I would really like to know what features the model is learning. https://pytorch-geometric.readthedocs.io/en/latest/tutorial/explain.html

Thanks, Sam

**Reply by rusty1s** [reply]
Yes, that's possible as long as your `edge_index` is part of the input.

**Comment by davodogster** [comment]
Hi @rusty1s 

This is the top down view of the Tree Point Cloud with KNNGraph used to create the edges. Do you think it's useful for learning tree params or too complex and a terrible representation?

![image](https://user-images.githubusercontent.com/46079516/227410827-118e8ea5-0e88-4774-bc16-f27e3487c463.png)

**Reply by rusty1s** [reply]
This is hard to tell from the image, but it looks reasonable, doesn't it?

**Comment by jediofgever** [comment]
@davodogster  Hi!,
Could you please share some of your insights and results? were you able to get decent results for multitask regression based on the clouds?

**Reply by davodogster** [reply]
Hi @jediofgever yes, so far the results have been quite promising for the PCD-KNN Graph Multitask Regression. Try different values of K for the KNN Graph like 5, 10, 20. Also try different architectures like PointNet vs GCN. Next we will try a more challenging dataset.

**Reply by jediofgever** [reply]
Thanks @davodogster, happy to hear that! I am at the very beginning, so I do not have the code yet, could you please share your repo for my reference? 
Also, are there any papers describing your work?
